### ISSUE TITLE
Numerical problems with torch.nn.functional.kl_div

### ISSUE THREAD

**Issue by superRookie007** [pytorch/pytorch#32520]
## 🐛 Bug
The implementation of **kl_div** in **torch.nn.functional** seems to have numerical problems. It can output **negative kl divergence**, and **small numbers even though they should be zero**. The problem could be due to the fact that it accepts input as **log_probability** while the target as **probability**. 

## To Reproduce
Suppose we have two outputs of the same values:
`a = torch.tensor([[1.0,2,3], [5.0, 5.0, 5.0]])`
`b = torch.tensor([[1.0,2,3], [5.0, 5.0, 5.0]])`

The output of the kl divergence according to the Pytorch implementation is:
`F.kl_div(F.log_softmax(a, 1), F.softmax(b, 1), reduction="none")`
`tensor([[-6.8829e-09, -1.4331e-08, -6.4222e-08],
        [ 1.6545e-08,  1.6545e-08,  1.6545e-08]])`

## Expected behavior
A simple implementation like below gives the correct behaviour, although probably not a very optimised implementation. 
`F.softmax(b, dim=1) * (F.log_softmax(b, dim=1) - F.log_softmax(a, dim=1))`
`tensor([[0., 0., 0.],
        [0., 0., 0.]])`

cc @ezyang @gchanan @zou3519

**Comment by albanD** [comment]
Thanks for the report. It looks like a numerical stability problem.
Maybe the same fix as the one we are currently doing for cosine_similarity (https://github.com/pytorch/pytorch/issues/29442) would be good here? Namely clamping the result to make sure it is greater than 0.

**Comment by ezyang** [comment]
cc @nikitaved

**Comment by rgommers** [comment]
@v0dro I unassigned you until the `triage review` label is cleared.

**Comment by ezyang** [comment]
Unilaterally removing this one from triage review

**Comment by nikitaved** [comment]
Interesting, the original code and the proposed fix are logically identical. Looks like there is a difference between `log(softmax)` and `log_softmax` implementations. Will have a look.

**Comment by gchanan** [comment]
@v0dro any update?

**Comment by rgommers** [comment]
@gchanan on the open PR for this I commented (https://github.com/pytorch/pytorch/pull/32616#discussion_r371430430) that since this is so similar to the `cosine_similarity` issue in gh-29442, to resolve the right approach there and then copy it for `kl_div`.

In that `cosine_similarity` issue, the preference was for a nicer but more complicated fix, and it breaks backwards compatibility to some extent (see https://github.com/pytorch/pytorch/pull/31378#issuecomment-578751908). That PR needs a decision from a maintainer.

If somehow `kl_div` is more urgent, you could consider merging the ad-hoc fix in gh-32616 until the `cosine_similarity` PR is decided on.

**Comment by nikitaved** [comment]
@rgommers, I am not sure that two issues are actually related. And seem like the real issue is in the `log_softmax`, not the `kl_div` itself. KL is computed in the most numerically-stable manner. Indeed, a bit confusing that one of the inputs represents its weights in a log-space, but, I guess, it is done to accommodate the `softmax` operation...

**Comment by nikitaved** [comment]
Hmm, `softmax` is cool. It does the scaling to improve its numerical stability.

**Comment by nikitaved** [comment]
Ok, it seem I know where the issue is. The issue is `log(softmax)`. Computing `log(softmax)` in a straightforward fashion in not numerically stable, we can rescale numerator and denominator by any exponent. `softmax/log_softmax` perform this rescaling...

**Comment by ezyang** [comment]
OK, we agreed with the proposed approach in our research meeting. Does this sufficiently unblock?

**Comment by nikitaved** [comment]
So, we agreed on an additional flag. This will affect the backward formula for `kl_div` as well.
Yeah, I believe it could be resolved already and it is not going to be bc-breaking.

**Comment by rgommers** [comment]
Done in commit 35cdb785228b8a, closing.

**Comment by Griffintaur** [comment]
Is the fix available in  version 1.5 or it will be part of next release 1.6

**Comment by ezyang** [comment]
```
ezyang-mbp:pytorch-tmp ezyang$ git log --pretty=format:%H v1.5.1 | grep 35cdb785228b8abea1d3bdb844aa5980e6642f8d
ezyang-mbp:pytorch-tmp ezyang$ git log --pretty=format:%H v1.6.0-rc2 | grep 35cdb785228b8abea1d3bdb844aa5980e6642f8d
35cdb785228b8abea1d3bdb844aa5980e6642f8d
```

It will be part of next release 1.6
